### PR TITLE
TA Plugin

### LINKED ISSUE #159: Register TA via bot
Make TA role no longer Mentionable. Instead, moderator can give TA role via bot.

```
!ta add User#0000 cop0000_asdf
``` 
Similar functionality for remove. Either remove all or remove for one class. When we delete the class channels we can reset the TA registrations too.

A student can use

```
!ta ask <some question>
```

And the bot will automatically tag the right people with the question. 

---
This way, if you're a TA in one class and a student in the other, you won't get tagged in both classes.

### PR DESCRIPTION
**Pull request by tanndlin**
fixes #159 
Still need to refactor some, but it works as is.

### CODE REVIEW

File: src/app/plugins/ta.plugin.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  private _handleAsk(message: IMessage, question: string) {
+    const channelTopic = (<TextChannel>message.channel).topic || '';
+    const hasTA: Boolean = channelTopic.indexOf('TA: ') !== -1;
+
+    if (!hasTA) {
+      message.reply('There are no TAs registered in this channel');
+      return;
+    }
+
+    const TA_tags = channelTopic.split('| TA: ')[1].split(' ');
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:76]: how does this work if someone has a space in their name?

File: src/app/plugins/ta.plugin.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      });
+
+    message.channel.send(`${message.author} asks: \n> ${question}\n${mentions}`);
+  }
+
+  private _handleRegister(message: IMessage) {
+    const channel: GuildChannel = message.channel as GuildChannel;
+    const channelTopic = (channel as TextChannel).topic || '';
+
+    const hasTA = Boolean(channelTopic.indexOf('TA: ') !== -1);
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:92]: `Boolean` construct not necessary

File: src/app/plugins/ta.plugin.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    if (existingTAs.split(' ').includes(message.author.tag)) {
+      message.reply(`You are already registered as a TA for ${channel.name}.`);
+      return;
+    }
+
+    channel
+      .setTopic(`${channelTopic} ${message.author.tag}`)
+      .then((newChan) => message.reply(`Successfully registered as TA in ${newChan.name}`));
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:112]: consider awaiting this and then awaiting the call to `handle...` in the `execute` routine

File: src/app/plugins/ta.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+  private _removeElement(arr: string[], toRemove: string): string[] {
+    const newArr: string[] = [];
+    arr.forEach((e) => {
+      if (e !== toRemove) {
+        newArr.push(e);
+      }
+    });
+    return newArr;
+  }
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:152]: you can also just do this with a `filter` and re-assign

File: src/services/channel.service.ts [hunk 5]
```diff
@@ -55,7 +55,7 @@ export class ChannelService {
     } else if (minimumChannelPermission === ChannelType.Bot) {
```
Comment (c650) [inline src/services/channel.service.ts:55]: nit: put this change in separate PR so we have a stamp of when this specific, very different and general, thing was changed

**Review by c650 — COMMENTED** [review]
nice code

File: src/app/plugins/ta.plugin.ts (hunk 1 above)
Comment (tanndlin) [reply src/app/plugins/ta.plugin.ts:76]: My bad, didn't know tags could have spaces

File: src/app/plugins/ta.plugin.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
+import { Plugin } from '../../common/plugin';
+import { IContainer, IMessage, ChannelType, Maybe } from '../../common/types';
+import { GuildChannel, GuildMember, TextChannel, Role } from 'discord.js';
+
+export class TaPlugin extends Plugin {
+  public name: string = 'TA Plugin';
+  public description: string = 'Allows TAs to register for classes.';
+  public usage: string = 'ta register <class_name> | ta ask <question>';
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:8]: class_name shouldn't be necessary; we can have the TA register within the class channel and then infer it from message.channel

File: src/app/plugins/ta.plugin.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  private _TA_ROLE = 'Teaching Assistant';
+  private _roleCache: Record<string, Maybe<Role>> = {
+    [this._TA_ROLE]: undefined,
+  };
+
+  constructor(public container: IContainer) {
+    super();
+  }
+
+  public validate(message: IMessage, args: string[]) {
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:21]: check if channel is a class channel, not just a private channel

File: src/app/plugins/ta.plugin.ts (hunk 6 above)
Comment (c650) [reply src/app/plugins/ta.plugin.ts:8]: yea i see that's what you do... so just change this help message by removing <class_name>

File: src/app/plugins/ta.plugin.ts [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+      .then((newChan) => message.reply(`Successfully unregistered as TA in ${newChan.name}`));
+  }
+
+  private _parseTags(list: string): string[] {
+    let temp = list;
+    const tags: string[] = [];
+    while (temp.length > 0) {
+      const index = temp.indexOf('#');
+      const name = temp.slice(0, index);
+      const discriminator = temp.slice(index, index + 5);
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:150]: +5 is magical, in other plugins you'll see I do a regex https://github.com/joey-colon/lion/blob/46e0f51ab44f39c2fe0acc230360b622c3639574/src/app/plugins/modreport.plugin.ts#L14-L28

File: src/app/plugins/ta.plugin.ts (hunk 6 above)
Comment (tanndlin) [reply src/app/plugins/ta.plugin.ts:8]: Sorry, old usage. I changed the way it registered TAs

File: src/app/plugins/ta.plugin.ts [hunk 9]
```diff
@@ -5,10 +5,11 @@ import { GuildChannel, GuildMember, TextChannel, Role } from 'discord.js';
 export class TaPlugin extends Plugin {
   public name: string = 'TA Plugin';
   public description: string = 'Allows TAs to register for classes.';
-  public usage: string = 'ta register <class_name> | ta ask <question>';
+  public usage: string = 'ta <register/remove> | !ta ask <question>';
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:8]: nit: `!` is implied

File: src/app/plugins/ta.plugin.ts [hunk 10]
```diff
@@ -5,10 +5,11 @@ import { GuildChannel, GuildMember, TextChannel, Role } from 'discord.js';
...
   public pluginAlias = [];
   public permission: ChannelType = ChannelType.Private;
 
+  private _DISCRIMINATOR_LENGTH: number = 5;
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:12]: would rather have it be `'#0000'.length`

File: src/app/plugins/ta.plugin.ts [hunk 11]
```diff
@@ -34,6 +35,13 @@ export class TaPlugin extends Plugin {
         .first();
     }
 
+    const isClassChannel = this.container.classService._isClassChannel(
+      (<TextChannel>message.channel).name
+    );
+    if (!isClassChannel) {
+      message.reply('Use this command in a class channel.');
```
Comment (c650) [inline src/app/plugins/ta.plugin.ts:42]: `return`

**Review by c650 — APPROVED** [review]
gg

**Review by c650 — APPROVED** [review]

**Comment by c650** [comment]
we gonna merge it?

File: src/services/class.service.ts [hunk 12]
```diff
@@ -173,6 +173,10 @@ export class ClassService {
     return responses;
   }
 
+  public _isClassChannel(className: string): Boolean {
```
Comment (joey-colon) [inline src/services/class.service.ts:176]: remove `_`, we only prepend private methods with them

File: src/app/plugins/ta.plugin.ts [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+      await this._handleRegister(message);
+      return;
+    }
+
+    if (subcommand === 'remove') {
+      await this._handleRemove(message);
+      return;
+    }
+
+    message.reply('Invalid sub-command');
```
Comment (joey-colon) [inline src/app/plugins/ta.plugin.ts:73]: might be useful to let them know what valid cmds are

File: src/app/plugins/ta.plugin.ts [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+      return;
+    }
+
+    const [subcommand, ...question] = args;
+    if (subcommand === 'ask') {
+      this._handleAsk(message, question.join(' '));
+      return;
+    }
+
+    const hasRole = member.roles.array().some((r) => r === this._roleCache[this._TA_ROLE]);
```
Comment (joey-colon) [inline src/app/plugins/ta.plugin.ts:57]: hasRole of what? `isTeachingAssistant` describes more

File: src/app/plugins/ta.plugin.ts [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+  constructor(public container: IContainer) {
+    super();
+  }
+
+  public validate(message: IMessage, args: string[]) {
+    return !!args.length;
+  }
+
+  public async execute(message: IMessage, args?: string[]) {
+    if (!args?.length) {
```
Comment (joey-colon) [inline src/app/plugins/ta.plugin.ts:27]: we already know args is valid because of line 23

File: src/app/plugins/ta.plugin.ts [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+      .members.filter((member) => TA_tags.some((TA: string) => member.user.tag === TA))
+      .forEach((TA: GuildMember) => {
+        mentions += `${TA.user} `;
+      });
+
+    message.channel.send(`${message.author} asks: \n> ${question}\n${mentions}`);
+  }
+
+  private async _handleRegister(message: IMessage) {
+    const channel: GuildChannel = message.channel as GuildChannel;
```
Comment (joey-colon) [inline src/app/plugins/ta.plugin.ts:95]: can we not immediately cast this to be a `TextChannel`?

File: src/app/plugins/ta.plugin.ts [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+    await channel
+      .setTopic(`${originalTopic} | TA: ${newTAs.join(' ')}`)
+      .then((newChan) => message.reply(`Successfully unregistered as TA in ${newChan.name}`));
+  }
+
+  private _parseTags(list: string): string[] {
+    let temp = list;
+    const tags: string[] = [];
+    while (temp.length > 0) {
```
Comment (joey-colon) [inline src/app/plugins/ta.plugin.ts:153]: could have just used a comma as delimiter and this wouldve been a lot simpler to implement

**Review by joey-colon — APPROVED** [review]

File: src/app/plugins/ta.plugin.ts (hunk 15 above)
Comment (tanndlin) [reply src/app/plugins/ta.plugin.ts:27]: it thinks args is `string[] | undefined` inside of execute without that

Comment (tanndlin) [reply src/app/plugins/ta.plugin.ts:27]: `if(!args){`?

File: src/app/plugins/ta.plugin.ts (hunk 17 above)
Comment (tanndlin) [reply src/app/plugins/ta.plugin.ts:153]: Should I leave as is, or change it to that now that I've done it the hard way  😆 
I don't mind, for the sake of better code

Comment (c650) [reply src/app/plugins/ta.plugin.ts:153]: can't comma appear in a name?

just checked: you can
